### PR TITLE
2409 remaining smart rollup operation types

### PR DESCRIPTION
**Pull request by zainen**
closes - #2409
Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [ ] Your code builds cleanly without any errors or warnings
- [ ] You have run the linter against the changes
- [ ] You have added unit tests (if relevant/appropriate)
- [ ] You have added integration tests (if relevant/appropriate)
- [ ] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

In this PR, please also make sure: 

- [ ] You have linked this PR to the issue by putting `closes #TICKETNUMBER` in the description box (when applicable)
- [ ] You have added a concise description on your changes
## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

File: packages/taquito-rpc/src/types.ts [hunk 1]
```diff
@@ -1445,6 +1602,45 @@ export interface OperationResultSmartRollupExecuteOutboxMessage {
   errors?: TezosGenericOperationError[];
 }
 
+export interface OperationResultSmartRollupPublish {
+  status: OperationResultStatusEnum;
+  consumed_milligas?: string;
+  staked_hash: string;
+  published_at_level: number;
+  balance_updates: OperationBalanceUpdates;
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1610]: these all should be optional except for status

File: packages/taquito-rpc/src/types.ts [hunk 2]
```diff
@@ -1445,6 +1602,45 @@ export interface OperationResultSmartRollupExecuteOutboxMessage {
...
+  errors?: TezosGenericOperationError[];
+}
+
+export interface OperationResultSmartRollupCement {
+  status: OperationResultStatusEnum;
+  consumed_milligas?: string;
+  inbox_level: number;
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1617]: same comment

File: packages/taquito-rpc/src/types.ts [hunk 3]
```diff
@@ -1445,6 +1602,45 @@ export interface OperationResultSmartRollupExecuteOutboxMessage {
...
+  errors?: TezosGenericOperationError[];
+}
+
+export interface OperationResultSmartRollupRefute {
+  status: OperationResultStatusEnum;
+  consumed_milligas?: string;
+  game_status: SmartRollupGameStatus;
+  balance_updates: OperationBalanceUpdates;
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1625]: same comment

File: packages/taquito-rpc/src/types.ts [hunk 4]
```diff
@@ -1445,6 +1602,45 @@ export interface OperationResultSmartRollupExecuteOutboxMessage {
...
+  errors?: TezosGenericOperationError[];
+}
+
+export interface OperationResultSmartRollupRecoverBond {
+  status: OperationResultStatusEnum;
+  balance_updates: OperationBalanceUpdates;
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1631]: this too

File: packages/taquito-rpc/src/types.ts [hunk 5]
```diff
@@ -1445,6 +1602,45 @@ export interface OperationResultSmartRollupExecuteOutboxMessage {
...
+  consumed_milligas?: string;
+  errors?: TezosGenericOperationError[];
+}
+
+export interface OperationResultSmartRollupTimeout {
+  status: OperationResultStatusEnum;
+  consumed_milligas?: string;
+  game_status: SmartRollupGameStatus;
+  balance_updates: OperationBalanceUpdates;
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1640]: same comment

File: packages/taquito-rpc/test/data/rpc-responses.ts [hunk 6]
```diff
@@ -5769,3 +5769,410 @@ export const smartRollupExecuteOutboxMessageResponse = {
     ],
   ],
 };
+
+export const smartRollupCementSample = {
```
Comment (dsawali) [inline packages/taquito-rpc/test/data/rpc-responses.ts:5773]: could you rename these to `smartRollupCementResponse` like the other examples for consistency? 

same with the other additional smart rollup operations

**Review by dsawali — CHANGES_REQUESTED** [review]

**Comment by dsawali** [comment]
could you link the issue to the PR please

File: packages/taquito-rpc/test/taquito-rpc.spec.ts [hunk 7]
```diff
@@ -4381,6 +4396,212 @@ describe('RpcClient test', () => {
     });
   });
 
+  describe('smartRollupPublish', () => {
+    it('should have correct types to access smart_rollup_publish results', async (done) => {
+      httpBackend.createRequest.mockReturnValue(Promise.resolve(smartRollupPublishResponse));
+      const response = await client.getBlock();
+      const content = response.operations[0][0]
+        .contents[0] as OperationContentsAndResultSmartRollupPublish;
```
Comment (AlirezaHaghshenas) [inline packages/taquito-rpc/test/taquito-rpc.spec.ts:4404]: Regarding the conversation we had about avoiding casting, we can modify this code to make sure that type narrowing works well for our types and users can avoid casting.

```
const content = response.operations[0][0].contents[0]; // no casting here
// test fields that are common here
if (content.kind !== OpKind.SMART_ROLLUP_PUBLISH) {
  fail('Some message');
  return; //maybe this is not needed.
}
// test fields that are only specific to OperationContentsAndResultSmartRollupPublish
```

**Review by dsawali — APPROVED** [review]

**Pull request merged**